Updating unit tests to be more flexible for warnings.

Katie-Howard opened the pull request:
Addresses #701 

Description

Testing:
 - [x] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

gavinevans commented on lib/improver/tests/ensemble_calibration/ensemble_calibration/test_ApplyCoefficientsForEnsembleCalibration.py:
```diff
@@ -887,11 +887,12 @@ def test_missing_date_catch_warning(self, warning_list=None):
             predictor_cube, variance_cube, optimised_coeffs,
             self.coeff_names, predictor_of_mean_flag)
 
+        warning_msg = "Ensemble calibration not available"
```
I think as well as modifying our test, so that we are only checking for the presence of a warning, rather than whether the warning is at a particular position within the warning list. we also want to remove the true on the line below that counts the number of warnings i.e.
```python
self.assertTrue(len(warning_list) == 1)
```
This will mean that the test won't fail, if the number of warnings increase e.g. due to changes in stacks. This affects a number of the other changes, where this line that counts the number of warnings can just be removed.

gavinevans commented on lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:
```diff
@@ -832,11 +836,13 @@ def test_current_forecast_cubes_len_zero_catch_warning(self,
 
         plugin.estimate_coefficients_for_ngr(
             current_forecast, historic_forecasts, truth)
+        warning_msg = "Insufficient input data present to estimate "
```
I think this warning message can be changed to:
```python
warning_msg = ("Insufficient input data present to estimate "
                           "coefficients using NGR.")
```
This is so the second line is included within the test of the warning message.

gavinevans commented on lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:
```diff
@@ -858,11 +864,13 @@ def test_historic_forecast_cubes_len_zero_catch_warning(self,
 
         plugin.estimate_coefficients_for_ngr(
             current_forecast, historic_forecasts, truth)
+        warning_msg = "Insufficient input data present to estimate "
```
Similar to above, brackets can be used for multi-line warning messages.

gavinevans commented on lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:
```diff
@@ -883,11 +891,13 @@ def test_truth_data_length_zero_catch_warning(self, warning_list=None):
 
         plugin.estimate_coefficients_for_ngr(
             current_forecast, historic_forecasts, truth)
+        warning_msg = "Insufficient input data present to estimate "
```
Similar to above, brackets can be used for multi-line warning messages.

Katie-Howard commented:
I think I have changed all the warnings to be more flexible now, and also removed the number count on the warnings where appropriate. So this should be completed now @gavinevans

gavinevans commented on lib/improver/tests/nowcasting/optical_flow/test_AdvectField.py:
```diff
@@ -343,9 +343,10 @@ def test_unmasked_nans(self, warning_list=None):
         expected_data[2, 1:] = np.array([2., 3.])
         result = self.plugin.process(cube, self.timestep)
-        self.assertTrue(len(warning_list) == 1)
+        warning_msg = "contains unmasked NaNs"
```
For the line below, where we are checking a particular position for a `UserWarning`, this should probably be replaced with the following, as is done in some of the other tests:
```python
        self.assertTrue(any(item.category == UserWarning
                            for item in warning_list))
```

Katie-Howard commented:
I have made those changes, so the tests should be flexible now. I have also done a slide and put it in the folder.
